Add metrics API scaler docs (#239)
kedacore/keda-docs: content/docs/2.0/scalers/metrics-api.md

+++
title = "Metrics API"
layout = "scaler"
availability = "v2.0+"
maintainer = "Community"
description = "Scale applications based on a metric from an API"
go_file = "metrics_api_scaler"
+++

### Trigger Specification

This specification describes the `metrics-api` trigger that scales based on a metric value from an API.

The metrics API requires following configuration values:

> - `url` string representing the base URL of the API (i.e `http://app:1317/api/v1/`).

Reviewer: ```suggestion
- `url` string representing the base URL of the API (eg. `http://app:1317/api/v1/`).
```

+++
title = "Metrics API"
layout = "scaler"
availability = "v2.0+"
maintainer = "Community"
description = "Scale applications based on a metric from an API"
go_file = "metrics_api_scaler"
+++

### Trigger Specification

This specification describes the `metrics-api` trigger that scales based on a metric value from an API.

The metrics API requires following configuration values:
- `apiURL` string representing the base URL of the API (i.e `http://app:1317/api/v1/`).
- `metricName` the metric name to be used to retive value .
- `targetValue` the target value is the target value to scale on. When the metric provided by the 
    API is equal or higher to this value, KEDA will start scaling out.

An example of scaled object spec using metric API:

```yaml

> spec:

Reviewer: I think you can remove the whole `spec:` section and start from `triggers`

+++
title = "Metrics API"
layout = "scaler"
availability = "v2.0+"
maintainer = "Community"

> description = "Scale applications based on a metric from an API"

Reviewer 2: ```suggestion
description = "Scale applications based on a metric provided by an API"
```

go_file = "metrics_api_scaler"
+++

### Trigger Specification

> This specification describes the `metrics-api` trigger that scales based on a metric value from an API. This scaler

Reviewer 2: ```suggestion
This specification describes the `metrics-api` trigger that scales based on a metric value provided by an API. This scaler
```

> allows users to utilize any existing APIs as a source of metric.  

Reviewer 2: ```suggestion
allows users to utilize any existing APIs as a metric source.  
```

Alternatively:
```suggestion
allows users to utilize any existing APIs as a metric provider.  
```

> The metrics-api trigger requires following configuration values:

Reviewer 2: ```suggestion
**Parameter list:**
```

> - `url` string representing the full URL that will be used to get the metric (eg. `http://app:1317/api/v1/stats`).

Reviewer 2: ```suggestion
- `url`: Full URL of the API operation to call to get the metric value (eg. `http://app:1317/api/v1/stats`).
```

> - `valueLocation` string using [GJSON path notation](https://github.com/tidwall/gjson#path-syntax) to point to

Reviewer 2: How about this?
```suggestion
- `valueLocation`: [GJSON path notation](https://github.com/tidwall/gjson#path-syntax) to refer to the field in the payload containing the metric value
```

    required value that will be used as current metric value 

> - `targetValue` the target value is the target value to scale on. When the metric provided by the 

Reviewer 2: ```suggestion
- `targetValue`: Target value to scale on. When the metric provided by the 
```

>     API is equal or higher to this value, KEDA will start scaling out.

Reviewer 2: Might be good to document this here already
```suggestion
    API is equal or higher to this value, KEDA will start scaling out. When the metric is 0 or less, KEDA will scale down to 0.
```

Here is an example of trigger configuration using metric-api scaler:

```yaml
triggers:
- type: metric-api
  metadata:
    targetValue: "8"
    url: "http://api:3232/api/v1/stats"
    valueLocation: "components.worker.tasks"
```

When checking current metric this scaler sends GET request to provided `url` and then uses `valueLocation`

> to access the value in response's payload. The above example expects that the API endpoint will return response

Reviewer 2: To improve rendered docs:
```suggestion
to access the value in response's payload.

The above example expects that the API endpoint will return response
```

similar to this one:
```json
{
  "components": {
    "worker": {
      "tasks": 12,
      ...
    },
    ...
  },
  ...
}
```

> Assuming such response, metric-api trigger will figure out that current metric value is 12.

Reviewer 2: ```suggestion
Assuming such response, Metrics API trigger will figure out that current metric value is 12.
```
or:
```suggestion
Assuming such response, `metric-api` trigger will figure out that current metric value is 12.
```

> **Note**:

Reviewer 2: Might be good to align with our other notes:
```suggestion
> **Note**: This scaler scales deployment to 0 if and only if the value of the metric is lower or equal to zero.
```
Similar to https://github.com/kedacore/keda-docs/blob/master/content/docs/2.0/concepts/scaling-deployments.md#L193

Alternatively, use `notice` as we do for Azure Event Hubs but not a must.
https://github.com/kedacore/keda-docs/blob/master/content/docs/2.0/scalers/azure-event-hub.md

+++
title = "Metrics API"
layout = "scaler"
availability = "v2.0+"
maintainer = "Community"
description = "Scale applications based on a metric provided by an API"
go_file = "metrics_api_scaler"
+++

### Trigger Specification

This specification describes the `metrics-api` trigger that scales based on a metric value provided by an API. 

> This scaler allows users to utilize any existing APIs as a metric provider.  

Reviewer 2: Just a suggestion to improve rendered docs:
```suggestion

This scaler allows users to utilize any existing APIs as a metric provider.  
```

**Parameter list:**
- `url`: Full URL of the API operation to call to get the metric value (eg. `http://app:1317/api/v1/stats`).
- `valueLocation`: [GJSON path notation](https://github.com/tidwall/gjson#path-syntax) to refer to the field in 
    the payload containing the metric value
- `targetValue`: Target value to scale on. When the metric provided by the API is equal or higher to this value, 
    KEDA will start scaling out. When the metric is 0 or less, KEDA will scale down to 0.

Here is an example of trigger configuration using metric-api scaler:

> ```yaml

Reviewer 2: Nit - This is typically listed before the parameters list; it's easier to resonate.

triggers:
- type: metric-api
  metadata:
    targetValue: "8"
    url: "http://api:3232/api/v1/stats"
    valueLocation: "components.worker.tasks"
```

### Authentication Parameters

Not supported yet.

> ### Example

Reviewer 2: Here we typically add a full YAML sample on how to use it.

Author: Added, used an example that I was using for testing purposes

When checking current metric this scaler sends GET request to provided `url` and then uses `valueLocation`
to access the value in response's payload. 

The above example expects that the API endpoint will return response similar to this one:
```json
{
  "components": {
    "worker": {
      "tasks": 12,
      ...
    },
    ...
  },
  ...
}
```
Assuming such response, Metrics API trigger will figure out that current metric value is 12.

> > NOTE: This scaler scales deployment to 0 if and only if the value of the metric is lower or equal to zero.

Reviewer 2: Actually I think we can remove it since it's in the parameters list

+++
title = "Metrics API"
layout = "scaler"
availability = "v2.0+"
maintainer = "Community"
description = "Scale applications based on a metric provided by an API"
go_file = "metrics_api_scaler"
+++

### Trigger Specification

This specification describes the `metrics-api` trigger that scales based on a metric value provided by an API. 

> This scaler allows users to utilize **any existing APIs** as a metric provider.  

Author: Bold to emphasize than any API mean in fact any API, so one can scale their deployment using current day length from https://ipgeolocation.io/astronomy-api.html 😃

> +++

Reviewer 2: Would be good to follow same structure from other scalers such as https://github.com/kedacore/keda-docs/blob/master/content/docs/2.0/scalers/redis-streams.md.

Let's talk about the Trigger Specification first, add a section on the API they need and what needs to comply with and end with an example of how to configure it.

title = "Metrics API"
layout = "scaler"
availability = "v2.0+"
maintainer = "Community"

> description = "Scale applications based on a metric value from an API"

Reviewer 2: This feels more natural to me but that's personal pref, and a big nit😂
```suggestion
description = "Scale applications based on a metric from an API"
```

But not a must though!

go_file = "metrics_api_scaler"
+++

This specification describes the `metrics-api` trigger that scales based on a metric value from an API.

### Trigger prerequisites

To use an external API as a source of metric you need to have a following endpoint:

> ```yaml

Reviewer 2: Let's add the whole OpenAPI spec here so that people can use it to validate their endpoint

  get:
    summary: Get Metric
    description: Get value for a given metric
    operationId: GetMetric
    parameters: 
     - name: metric
       in: path
       required: true
       schema:
          type: string
       description: Name of the metric for which information is requested
    responses:
      '200':
        description: Metric inforation was provided
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/ReportedMetric'
      '404':
        description: Metric is not supported
      '401':
        $ref: '#/components/responses/UnauthorizedError'
      '500':
        description: Unable to determine metric information
    tags:
      - Metrics
    security:
      - basic_auth: []
      - api_key: []
```
which returns:
```yaml
ReportedMetric:
  type: object
  required:
    - name
    - value
  properties:
    name:
      type: string
    value:
      type: number
```

### Trigger Specification

The metrics API requires following configuration values:
- `apiURL` string repesenting the base URL of the API (i.e `http://app:1317/api/v1/`).
- `metricName` the metric name to be used to retive value .

> - `targetValue` the target value for HPA. As long as the current metric doesn't match `TargetValue`, 

Reviewer 2: We cannot assume everybody knows what an HPA is (that's one of the selling points of KEDA, no need to know how Kubernetes scales), so how about:

```suggestion
- `targetValue` is the target value to scale on. When the metric provided by the API is equal or higher to this value, KEDA will start scaling out.
```

  HPA will increase the number of the pods until it reaches the maximum number of pods allowed to scale to.

An example of scaled object using metric API:

```yaml

> apiVersion: keda.k8s.io/v1alpha1

Reviewer: ```suggestion
apiVersion: keda.sh/v1alpha1
```

kind: ScaledObject
metadata:
  name: metrics-api-scaledobject
  namespace: default

>   labels:

Reviewer 2: No need to include this

    deploymentName: dummy
spec:
  maxReplicaCount: 4
  scaleTargetRef:

>     deploymentName: dummy

Reviewer: ```suggestion
    name: dummy
```

  triggers:
    - type: metric-api
      metadata:
        targetValue: "1"

>         apiURL: "http://api:3232/api/v1"

Reviewer 2: I would just go for `url` here since it's the metrics API scaler
```suggestion
        url: "http://api:3232/api/v1"
```

Reviewer 2: Another question - Should they configure `http://api:3232` or `http://api:3232/api/v1`? Given we've defined the operations and the endpoint might evolve over time I'd argue that it should just be `http://api:3232` and we'll handle the rest, no?

Reviewer 2: What do you think @zroubalik @ahmelsayed ?

Author: @tomkerkhove that means that we have to incorporate the API version somewhere else, right?

Reviewer 2: Given they have to comply with our OpenAPI spec it's already covered.

Reviewer 2: https://github.com/kedacore/keda-docs/pull/239/files#diff-2079ef1d5bc7dbe2c2470f43d8749c87R17

Reviewer 2: If that endpoint doesn't exist, then the API is not supported 🤷‍♂️.

If we introduce a `v2

Reviewer: I like the current approach, I am not sure if there's a benefit in splitting those fields (the config gets more complex)

Author: But if I update my API version in a backward-compatible manner then the endpoint will exist. I think it's better to have full base link as this allows more extensibility. For example, some may have `http://app:1234/experimental/api/v1/` due to ongoing migration process or whatever

Reviewer 2: I'm not saying it has to be separated but just that we care about the base URL.

All metric APIs need to comply with the OpenAPI spec; if we add endpoints there, KEDA will determine what the best one is to call. If other endpoints exist next to it, we'll simply ignore them.

If we configure the full prefix in it, what's the point of having an OpenAPI spec anyway :)

Author: >I'm not saying it has to be separated but just that we care about the base URL.

What you mean by "base URL"?

> If we configure the full prefix in it, what's the point of having an OpenAPI spec anyway :)

To be honest, that's the advantage. If you use OpenAPI spec that's good. If you don't... well you can still use the scaler. And in my opinion, that's an advantage. Moreover, we only use `api/v1//metric/{name}` in scaler

Reviewer 2: With base URL I was referring to `http://app:1234`, the rest is handled by the scaler which knows `api/v1/metric/{name}` must be there so it just adds it automatically.

Reviewer 2: For me every user has to use that endpoint, if you want to use ´http://app:1234/experimental/api/v1/` then that's a violation imo

Author: Btw. here's @zroubalik comment https://github.com/kedacore/keda-docs/pull/239#discussion_r477143636

If we use only `api/v1/metric/{name}` then how users can change the API version they want to use? And the version can change.

> if you want to use ´http://app:1234/experimental/api/v1/` then that's a violation imo

I would call this violation an "extensibility". Let assume that base url is `http://app:1234`. Should we validate it to do not allow `http://app:1234/staging/user/test/and/more/` as input? 😉
